command/repo.go

```diff
@@ -63,6 +83,163 @@ func repoClone(cmd *cobra.Command, args []string) error {
 	return utils.PrepareCmd(cloneCmd).Run()
 }
 
+func isURL(arg string) bool {
+	return strings.HasPrefix(arg, "http:/") || strings.HasPrefix(arg, "https:/")
+}
+
+var Since = func(t time.Time) time.Duration {
+	return time.Since(t)
+}
+
+func repoFork(cmd *cobra.Command, args []string) error {
+	ctx := contextForCommand(cmd)
+
+	clonePref, err := cmd.Flags().GetString("clone")
+	if err != nil {
+		return err
+	}
+	remotePref, err := cmd.Flags().GetString("remote")
+	if err != nil {
+		return err
+	}
+
+	apiClient, err := apiClientForContext(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to create client: %w", err)
+	}
+
+	var toFork ghrepo.Interface
+	inParent := false // whether or not we're forking the repo we're currently "in"
+	if len(args) == 0 {
+		baseRepo, err := determineBaseRepo(cmd, ctx)
+		if err != nil {
+			return fmt.Errorf("unable to determine base repository: %w", err)
+		}
+		inParent = true
+		toFork = baseRepo
+	} else {
+		repoArg := args[0]
+
+		if isURL(repoArg) {
+			parsedURL, err := url.Parse(repoArg)
+			if err != nil {
+				return fmt.Errorf("did not understand argument: %w", err)
+			}
+
+			toFork, err = ghrepo.FromURL(parsedURL)
+			if err != nil {
+				return fmt.Errorf("did not understand argument: %w", err)
+			}
+
+		} else {
+			toFork = ghrepo.FromFullName(repoArg)
+			if toFork.RepoName() == "" || toFork.RepoOwner() == "" {
+				return fmt.Errorf("could not parse owner or repo name from %s", repoArg)
+			}
+		}
+	}
+
+	greenCheck := utils.Green("✓")
+	redX := utils.Bold(utils.Red("X"))
+	out := colorableOut(cmd)
+	s := utils.Spinner()
+	loading := utils.Gray("Forking ") + utils.Bold(utils.Gray(ghrepo.FullName(toFork))) + utils.Gray("...")
+	s.Suffix = " " + loading
+	s.FinalMSG = utils.Gray(fmt.Sprintf("- %s\n", loading))
+	s.Start()
+
+	authLogin, err := ctx.AuthLogin()
+	if err != nil {
+		s.Stop()
+		return fmt.Errorf("could not determine current username: %w", err)
+	}
+
+	possibleFork := ghrepo.New(authLogin, toFork.RepoName())
+
+	forkedRepo, err := api.ForkRepo(apiClient, toFork)
+	if err != nil {
+		s.Stop()
+		return fmt.Errorf("failed to fork: %w", err)
+	}
+
+	// This is weird. There is not an efficient way to determine via the GitHub API whether or not a
+	// given user has forked a given repo. We noticed, also, that the create fork API endpoint just
+	// returns the fork repo data even if it already exists -- with no change in status code or
+	// anything. We thus check the created time to see if the repo is brand new or not; if it's not,
+	// we assume the fork already existed and report an error.
+	created_ago := Since(forkedRepo.CreatedAt)
+	if created_ago > time.Minute {
+		s.Stop()
+		fmt.Fprint(out, redX+" ")
+		return fmt.Errorf("%s already exists", ghrepo.FullName(possibleFork))
```

Comment:
If the fork already exists, couldn't we just proceed with the rest of this command's logic, but adjust the output accordingly

i.e. instead of "Created fork X" we could just print "Existing fork X found" and continue as normal?

Reply:
I decided against it since I felt like the fork already existing was indicative of a user doing something by accident; but thinking about it they might appreciate the second step (ie maybe they thought they hadn't forked it because they hadn't yet cloned or added a remote). I'll switch.